### DISCUSSION TITLE
SharpGen.Runtime.ComObject.Release

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#338] [Q&A]
When I call SharpGen.Runtime.ComObject.Release with..
```
uint refCount = DxDevice.Release();
```
Does that give all live objects on the device. That's the way I've been using it, but I just wanted to check because I'm going over my shutdown code and I want to be sure I understand it.

Edit: Actually, all I'm doing here is checking if the device itself has a reference count aren't I ?

Edit 2:  Actually, I don't know if I should be calling this where I do because I may be skipping my live objects reporting and device disposal. Can someone just comment on this please?
```
  Dispose(...) 
  {
          ...
          DxDeviceContext.ClearState();
          DxDeviceContext.Flush();
          DxDeviceContext.Release();
          DxDeviceContext.Dispose();
          DxDeviceContext = null;

          #if DEBUG
          uint refCount = DxDevice.Release();
          if (refCount > 0)
          {
              System.Diagnostics.Debug.WriteLine($"Direct3D11: There are {refCount} unreleased references left on the device");
              ID3D11Debug d3d11Debug = DxDevice.QueryInterfaceOrNull<ID3D11Debug>();
              if (d3d11Debug != null)
              {
                  ... 
                  d3d11Debug.ReportLiveDeviceObjects( .... );
                  d3d11InfoQueue.Dispose();
                  d3d11Debug.Dispose();
              }
              DxAdapter.Dispose();
```

**Comment by amerkoleci** [comment]
During debug you can call Release and check refCount of device and report leaks, or you can use DXGIGetDebugInterface1 as well, take look here: 

https://github.com/amerkoleci/Vortice.Windows.Samples/blob/main/src/Vortice.Framework/D3D11Application.cs#L165

During NON Debug build you can just Dispose objects.
